Make sure lazy engines with async routes are handled

jackson-dean opened the pull request:
I was running into this error when loading a lazy engine route:
```
TypeError: route.prefetch is not a function
```
`route` in this case was actually a promise which resolved to a route with prefetch method.

This converts the change set in the current iteration to a promise to ensure the correct fullParams are used and then normalizes the route to always be a promise so they are treated the same for both cases.

jackson-dean commented:
There is a failing test which I will look into if this is generally the right thing to do.

jackson-dean commented:
It looks like all tests pass but there is some weird issue where running yarn does not generate node_modules for the linked lazy-engine. https://travis-ci.org/nickiaconis/ember-prefetch/jobs/586819795

I also see this issue locally. however if I simply run yarn again everything works out fine. any ideas where I might be going wrong?

jackson-dean commented:
Ok, a little tinkering with the travis configs and I finally got everything working.

jackson-dean commented on addon/services/prefetch.js:
```diff
@@ -28,11 +28,20 @@ if (gte('3.6.0')) {
-                  });
-                  seenRoutes.set(route, true);
-                  if (transition.isAborted) return;
+                route._prefetched = new RSVP.Promise(r => {
```
Actually I realize looking at this now that is does not really make sense to set `_prefetched` on the promise...

jackson-dean replied:
Actually I realize looking at this now that is does not really make sense to set _prefetched in the cases where the route is a promise instance.

jackson-dean commented:
after some extra experimenting I'm not so sure this the right thing to do. but at least it is here to get feedback and start a discussion about it.

rwjblue commented on .gitignore:
```diff
@@ -5,7 +5,7 @@
 
 # dependencies
-/node_modules/
+**/node_modules/
```
I don't think we should end up with nested node_modules folders, what problem is this solving?

rwjblue commented on .travis.yml:
```diff
@@ -3,7 +3,7 @@ language: node_js
   # we recommend testing addons with the same minimum supported node version as Ember CLI
   # so that your addon works for all apps
-  - "6"
+  - "10"
```
Can you make this change in a separate PR? This should use Node 8 (not 10), as it is the minimum supported Node version. In addition, please make sure to also update the `package.json` `engines` field to show that Node 8 is the minimum supported version.

jackson-dean replied:
Sure! That does make sense, I think at the time I was just trying to get things working, but this certainly makes sense

rwjblue replied:
Ya, I definitely understood that. I just like keeping them separate when we actually go to land the changes (makes reviewing the logs easier, changelog generation, etc).

jackson-dean replied:
Added separate PR for updating CI configs https://github.com/nickiaconis/ember-prefetch/pull/151. That means this will be failing CI until that is merged, but to be expected and we can indeed evaluate the CI changes in isolation.

rwjblue commented on .travis.yml:
```diff
@@ -57,5 +57,8 @@ before_install:
   - npm --version
 
+install:
+  - yarn install --no-lockfile --non-interactive
```
This seems somewhat unrelated, I'd like to get CI passing without the runtime changes in ths PR so that we can evaluate things independently.

Also, it seems odd to use both `npm` and `yarn` in the same CI config. I would think that you should remove the various `npm` commands just above as well.

jackson-dean replied:
yes, fair assessment about npm. Tbh I don't completely understand why this makes things work, but it all has to do with the dummy engines being linked into node_modules. Just installing directly with npm (or yarn) without the flags throws an error about node_modules missing inside the engine folders. There could be a different way to address this, but again I just followed the example config in ember-engines repo to get things working https://github.com/ember-engines/ember-engines/blob/master/.travis.yml#L54

rwjblue replied:
`npm` doesn't support `link:` as a prefix. if you want to use `npm` you have to use `file:` which does a symlink (but under `yarn` using `file:` means _duplicate copy_ into `node_modules` 🤕 ).

I'm super happy to move to `yarn`, I just want to be _fully_ in one camp or the other (and preferably in a separate "use yarn for CI" PR).

jackson-dean replied:
Created a new PR for just the CI config changes: https://github.com/nickiaconis/ember-prefetch/pull/151

rwjblue commented on addon/services/prefetch.js:
```diff
@@ -25,14 +25,16 @@ if (gte('3.6.0')) {
+                // normalize route to always be a promise since lazy loaded
+                // routes are promises, while preloaded routes are not. We want
+                // to treat them both the same.
+                RSVP.resolve(changeSet.for[i]).then(({ route, fullParams }) => {
```
As far as I can tell, `changeSet.for` can never be a promise.

`createPrefetchChangeSet` does not use promises, and returns a synchronous value:

https://github.com/nickiaconis/ember-prefetch/blob/fc5a60281536129e57e240a0b6eb15c565382b3f/addon/-private/diff-route-info.js#L127-L157

The `changeSet.for` value is the result of calling `getPrefetched` here, which does not use promises either:

https://github.com/nickiaconis/ember-prefetch/blob/fc5a60281536129e57e240a0b6eb15c565382b3f/addon/-private/diff-route-info.js#L114-L125

jackson-dean replied:
yes, you are 100% right about that and it does look odd that I would do this, but I had some intuition that without this we could end up using an unexpected `fullParams` object inside the promise `then` handler since it pulls this out of changeSet.for synchronously in the loop before the callback runs. I could be mistaken here, but it seemed to me like there could be a situation where we end up using the `fullParams` from the last loop iteration when we don't really mean to. Or maybe my intuition was just wrong and I'm not thinking clearly about it.

jackson-dean replied:
I actually changed the implementation a bit to handle promise case and synchronous case separately. Let me know what you think.

jackson-dean replied:
I liked having a generic way of handling things, but I think the trade off there is that its not really clear why certain things are being done. I think the new approach makes it much more explicit what is happening and why.

jackson-dean replied:
made some other minor changes. I think we can use the promise object directly in seenRoutes WeekMap. I don't think it really matters what type of object is used there.

rwjblue commented on addon/services/prefetch.js:
```diff
@@ -25,14 +25,16 @@ if (gte('3.6.0')) {
+                // routes are promises, while preloaded routes are not. We want
+                // to treat them both the same.
+                RSVP.resolve(changeSet.for[i]).then(({ route, fullParams }) => {
+                  RSVP.resolve(route).then(resolvedRoute => {
```
We should use `RSVP.Promise.resolve()` not `RSVP.resolve()` here.

```suggestion
                  RSVP.Promise.resolve(route).then(resolvedRoute => {
```

jackson-dean commented on .gitignore:
```diff
@@ -6,6 +6,7 @@
 
 # dependencies
 /node_modules/
+tests/dummy/lib/**/node_modules
```
now that ember engines are symlinked to node_modules it ends up tracking node_modules underneath the dummy app folder. I changed this to be much more specific about exactly which node_modules we are excluding. For much of this setup I just followed whatever ember-engines repo is doing. I just checked and this is how it is handled there as well: https://github.com/ember-engines/ember-engines/blob/master/.gitignore#L17

rwjblue replied:
Gotcha! Thank you, great explanation and that makes perfect sense.

jackson-dean commented on addon/services/prefetch.js:
```diff
@@ -25,14 +25,16 @@ if (gte('3.6.0')) {
+                // normalize route to always be a promise since lazy loaded
+                // routes are promises, while preloaded routes are not. We want
+                // to treat them both the same.
+                RSVP.Promise.resolve(changeSet.for[i]).then(({ route, fullParams }) => {
```
It does seem odd that I am wrapping `changeSet.for[i]` in a promise when it never actually could be a promise. I had some concern, however, that we could end up using the last `fullParams` object inside the `then` handler for doing prefetch when we did not mean to if we don't make sure to capture both the route and the fullParams at the current iteration to pass into the `then` callback. My intuition about this may be wrong and I may just not be thinking through it clearly, but it seemed like this could happen. Or there could be a completely different way to handle the situation. Does it my thinking make sense? thoughts?

jackson-dean commented on addon/services/prefetch.js:
```diff
@@ -28,11 +28,22 @@ if (gte('3.6.0')) {
 
-                  route._prefetched = new RSVP.Promise(r => {
-                    return r(route.prefetch(fullParams, transition))
+                let isRoutePromiseLike = route && route.then instanceof Function && route.catch instanceof Function;
```
let me know if you have a preferred convention of checking if something is a promise

rwjblue replied:
I think I would do something like:

```js
function isThenable(obj) {
  if ((typeof obj === 'object' && obj !== null) || typeof obj === 'function') {
    return typeof obj.then === 'function';
  }

  return false;
}
```

rwjblue commented:
Ready for a rebase now that #151 is landed.

jackson-dean commented on addon/services/prefetch.js:
```diff
@@ -28,11 +29,20 @@ if (gte('3.6.0')) {
+                  // handler is executed for lazy loaded routes.
+                  RSVP.resolve(fullParams).then(capturedFullParams => {
+                    route.then(resolvedRoute => {
+                      route._prefetched = RSVP.Promise.resolve(resolvedRoute.prefetch(capturedFullParams, transition));
```
I am thinking also it probably makes sense to check isDestroying and isAborted in this callback, but I'll wait for your next review to see what your thoughts are.

jackson-dean commented:
@rwjblue Thanks for all of your previous feedback! bump, whenever you get a chance. This is actually a blocker for us atm. Thank you!!

rwjblue commented:
sorry for the delay, will try to re-review today...

jackson-dean commented:
ok cool, I also noticed a bug as soon as I looked at the code again so I updated it and added an additional test to make sure prefetch *actually* works for a routeable lazy engine. however, now there is some funky eslint issue... looking into it

jackson-dean commented:
it looks like `yarn lint:js` started failing after I added more files in the engine. it's not immediately obvious to me what is busted, but this is error I see when I run it locally:
```
Cannot find module '@ljharb/eslint-config'
Referenced from: /Users/jcdean/ws/ember-prefetch/tests/dummy/lib/lazy-engine/node_modules/broccoli-babel-transpiler/node_modules/resolve/.eslintrc
Error: Cannot find module '@ljharb/eslint-config'
Referenced from: /Users/jcdean/ws/ember-prefetch/tests/dummy/lib/lazy-engine/node_modules/broccoli-babel-transpiler/node_modules/resolve/.eslintrc
    at ModuleResolver.resolve (/Users/jcdean/ws/ember-prefetch/node_modules/eslint/lib/util/module-resolver.js:74:19)
    at resolve (/Users/jcdean/ws/ember-prefetch/node_modules/eslint/lib/config/config-file.js:479:28)
    at load (/Users/jcdean/ws/ember-prefetch/node_modules/eslint/lib/config/config-file.js:551:26)
    at configExtends.reduceRight (/Users/jcdean/ws/ember-prefetch/node_modules/eslint/lib/config/config-file.js:425:36)
    at Array.reduceRight (<anonymous>)
    at applyExtends (/Users/jcdean/ws/ember-prefetch/node_modules/eslint/lib/config/config-file.js:403:26)
    at loadFromDisk (/Users/jcdean/ws/ember-prefetch/node_modules/eslint/lib/config/config-file.js:523:22)
    at Object.load (/Users/jcdean/ws/ember-prefetch/node_modules/eslint/lib/config/config-file.js:559:20)
    at Config.getLocalConfigHierarchy (/Users/jcdean/ws/ember-prefetch/node_modules/eslint/lib/config.js:227:44)
    at Config.getConfigHierarchy (/Users/jcdean/ws/ember-prefetch/node_modules/eslint/lib/config.js:179:43)
```

jackson-dean commented:
however, I see a completely different error in the CI log:
```
ESLint: 4.19.1.
ESLint couldn't find the plugin "eslint-plugin-prettier". This can happen for a couple different reasons:
1. If ESLint is installed globally, then make sure eslint-plugin-prettier is also installed globally. A globally-installed ESLint cannot find a locally-installed plugin.
2. If ESLint is installed locally, then it's likely that the plugin isn't installed correctly. Try reinstalling by running the following:
    npm i eslint-plugin-prettier@latest --save-dev
If you still can't figure out the problem, please stop by https://gitter.im/eslint/eslint to chat with the team.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
The command "yarn lint:js" exited with 1.
```

jackson-dean commented:
well, i came across this https://github.com/ember-cli/rfcs/pull/121

jackson-dean commented:
created a separate PR for updating the linting packages https://github.com/nickiaconis/ember-prefetch/pull/154

jackson-dean commented:
I ran eslint with these changes on top of the eslint changes and confirmed that it runs successfully locally.

jackson-dean commented on addon/services/prefetch.js:
```diff
@@ -28,9 +29,24 @@ if (gte('3.6.0')) {
+                if (isThenable(route)) {
+                  // Ensure we use the expected fullParams when the promise then
+                  // handler is executed for lazy loaded routes.
+                  RSVP.Promise.resolve(fullParams).then(capturedFullParams => {
```
If we convert the wrapping for loop to a forEach, I think we can get rid of this funky resolve wrapping the full params

jackson-dean replied:
what do you think @rwjblue , do we really need to be able to break? seems there would not be a deep enough route hierarchy for it to really matter.

rwjblue commented:
Thank you for all the work here @jackson-dean! This was ultimately replaced by https://github.com/nickiaconis/ember-prefetch/pull/159 which was released in v4.0.1.

Please, let me know if there are other changes/fixes that you are still needing...

jackson-dean commented:
👍 I took a look. It does seem it will fix the issue of prefetch not getting called on a lazy engine, but it still seems we could end up passing unexpected transition and fullParams when the route resolves since those values are extracted in a loop outside of the promise callback. Whether it is actually a problem in practice I'm not sure. If not, it really does not matter that much.

xg-wang commented:
> when the route resolves since those values are extracted in a loop outside of the promise callback

Thanks @jackson-dean for pointing out! Is this the code you're referring to? https://github.com/tildeio/router.js/blob/v6.2.5/lib/router/route-info.ts#L240
This is a public API in ember https://api.emberjs.com/ember/3.13/classes/Route/methods/serialize?anchor=serialize.

However the serialization happens after waiting model hook `this.getModel(transition)`. It does not make sense for prefetch hook to wait model hook.

I think it’s fine to not wait the serialized params, but a bit documentation about https://api.emberjs.com/ember/3.13/classes/Route/methods/serialize?anchor=serialize would not be called for the `prefetch(params, transition)` and the `transition` tenable isn’t fully resolved yet.

jackson-dean commented:
Hey @xg-wang my concern was just misplaced. I was talking about here https://github.com/nickiaconis/ember-prefetch/pull/159/files#diff-54a7d68a8480ef36da94321be6541ca1R29 but my instinct was completely incorrect and just a consequence of working with js before `let` and `const` was even a thing 😅  , but just by virtue of using `let` which has proper scoping semantics, everything does indeed work out fine! If this were all `var` declarations it would be broken, but because js is more sane now it is perfectly fine 😄 

However, if this code is transpiled with babel to old `var` statements, I think there still could be a problem 🤔 Unless babel is somehow smart enough to resolve this inconsistency.

jackson-dean commented:
actually, on yet another look, it looks it will even be fine in any case because the outer promise immediately invokes its resolver with the outer route and fullParams. So, yes, at the end I don't have any concern about it. I had also initially missed this line https://github.com/nickiaconis/ember-prefetch/pull/159/files#diff-54a7d68a8480ef36da94321be6541ca1R23 which also mitigates any concern.

nice work, and thanks for fixing this! 😄
